model_api/python/openvino/model_api/models/anomaly.py

```diff
@@ -0,0 +1,154 @@
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+"""
+from typing import Any
+
+import cv2
+import numpy as np
+
+from .image_model import ImageModel
+from .types import ListValue, NumericalValue, StringValue
+from .utils import AnomalyResult
+
+
+class AnomalyDetection(ImageModel):
+    __model__ = "AnomalyDetection"
+
+    def __init__(self, inference_adapter, configuration=None, preload=False):
+        super().__init__(inference_adapter, configuration, preload)
+        self.output_name = self.inference_adapter.model.outputs[0].any_name
+        # attributes for mypy
+        self.max: float
+        self.min: float
+        self.image_threshold: float
+        self.pixel_threshold: float
+        self.task: str
+
+    def preprocess(self, inputs: np.ndarray):
+        inputs = inputs / 255.0  # model expects inputs in range [0, 1]
+        return super().preprocess(inputs)
+
+    def postprocess(self, outputs: dict[str, np.ndarray], meta: dict[str, Any]):
+        """Post-processes the outputs and returns the results.
+
+        Args:
+            outputs (dict[str, np.ndarray]): Raw model outputs
+            meta (dict[str, Any]): Meta data containing the original image shape
+
+        Returns:
+            AnomalyResult: Results
+        """
+        anomaly_map: np.ndarray | None = None
+        pred_label: str | None = None
+        pred_mask: np.ndarray | None = None
+        pred_boxes: np.ndarray | None = None
+        box_labels: np.ndarray | None = None
+
+        predictions = outputs[self.output_name]
+
+        if len(predictions.shape) == 1:
+            pred_score = predictions
+        else:
+            anomaly_map = predictions.squeeze()
+            pred_score = anomaly_map.reshape(-1).max()
+
+        if hasattr(self, "image_threshold"):
+            pred_label = "Anomalous" if pred_score > self.image_threshold else "Normal"
+
+        if hasattr(self, "task") and self.task in ("segmentation", "detection"):
+            assert anomaly_map is not None  # for mypy
+            pred_mask = (anomaly_map >= self.pixel_threshold).astype(np.uint8)
+            anomaly_map = self._normalize(anomaly_map, self.pixel_threshold)
+
+        # normalize
+        pred_score = self._normalize(pred_score, self.image_threshold)
+
+        # resize outputs
+        if anomaly_map is not None:
+            anomaly_map = cv2.resize(
+                anomaly_map, (meta["original_shape"][1], meta["original_shape"][0])
+            )
+            pred_mask = cv2.resize(
+                pred_mask, (meta["original_shape"][1], meta["original_shape"][0])
+            )
+
+        if hasattr(self, "task") and self.task == "detection":
+            pred_boxes = self._get_boxes(pred_mask)
+            box_labels = np.ones(pred_boxes.shape[0])
+
+        return AnomalyResult(
+            anomaly_map=anomaly_map,
```

Comment:
Does that task handles both detection and segmentation at once?

Reply:
Yes. Since bounding boxes are computed from the mask, the code just checks if the task type is detection and adds bounding boxes to the results as well.